### PR TITLE
Added RetinaNet example

### PR DESCRIPTION
**Pull request by srihari-humbarwadi**
I have added a script containing the retinanet example, can you please take a look. I will run the script to generate the other files and upload them once this is approved.

The script demonstrates building retinanet and training on a simple dataset, I couldn't demo training on ms coco because it relatively big and takes more than 30 mins to train each epoch.
The script also supports training on multiple GPUs and on cloud TPUs, implementing custom layers, overiding train, test and predict step functions and exporting a custom signature in the saved model. The notebook version of this takes about 20 mins to run on Colab GPU.

### CODE REVIEW

File: examples/vision/retinanet.py [hunk 1]
```diff
@@ -0,0 +1,1217 @@
...
+
+class RetinaNetClassificationLoss(tf.losses.Loss):
+    def __init__(self, alpha, gamma):
+        super(RetinaNetClassificationLoss, self).__init__(
+            reduction="none", name="RetinaNetClassificationLoss"
+        )
+        self._alpha = alpha
+        self._gamma = gamma
+
+    def _focal_loss(self, y_true, y_pred):
```
Comment (fchollet) [inline examples/vision/retinanet.py:669]: There is no point in making this method different from `call`.

**Comment by fchollet** [comment]
Thank you for the PR! I think a code example demonstrating RetinaNet will be very valuable.

There are a number of significant changes we would have to do before submitting this.

1. Length. A code example should be shorter than 300 lines of code (this doesn't include text & comments).
2. Complexity. The code can be significantly simplified.
3. Dataset. Please use MS COCO and check that you get industry-standard results on MS COCO. In the actual example we submit, we want to limit running time, so you can set the number of training epochs to only 1. If that's still too long, add a `ds.take(100)` line to restrict the dataset to the first 100 batches (while commenting that this line should be removed when readers attempt to train the model to convergence)
4. Code patterns. See below.

About code patterns:

- It is my impression that you've reused significant parts of the RetinaNet from the Model garden. This is not what you should do: the Model Garden model was ported from TF 1 and does not represent idiomatic Keras code. Our goal is to get to a model written from scratch in idiomatic Keras, which would look very different from the Model Garden example.
- You don't need all the distribution strategy-related code.
- You don't need all the loss scaling-related code.
- You don't need to override `compile()`.
- You shouldn't have a separate `self._network` model, the model you train should be the same object.

If you would like to keep working on it, please let me know. Otherwise, I plan on working on a RetinaNet implementation myself in 2 weeks (as part of a hackathon).

**Comment by srihari-humbarwadi** [comment]
Thank you for the detailed review, I would definitely want to work on this and make the  necessary changes. 
I haven't used code from the model garden, but yes i can definitely make the code simpler and short. The only thing I reused was Resnet50 from keras_applications. Should i implement it from scratch aswell?.

**Comment by fchollet** [comment]
> Thank you for the detailed review, I would definitely want to work on this and make the necessary changes.

Ok, sounds good, thank you!

> Almost the whole implementation is from scratch and the only thing I reused was Resnet50 from keras_applications. Should i implement it from scratch aswell?

No need, you should reuse as much built-in functionality as possible.

**Comment by srihari-humbarwadi** [comment]
I am trying to add the MS COCO dataset in a way that it requires the minimum amount of preprocessing code, so obviously, tensorflow_datasets is the best choice, but I noticed that the MS COCO dataset is pretty big, and even if we limit the number of training batches when we submit the example, the users would still have to download at least 19GB of data, which takes quite some time in some situations. So what would be the best way to go around this?

**Comment by fchollet** [comment]
I believe you should be able to use e.g. `split="train[0:1000]"` to only load the first 10,000 samples.

**Comment by fchollet** [comment]
After further checking, it turns out that avoiding a full download of COCO via TFDS is impossible. So we should either not use TFDS, or use a different (smaller) dataset than COCO.

**Comment by AakashKumarNain** [comment]
> It is my impression that you've reused significant parts of the RetinaNet from the Model garden. This is not what you should do: the Model Garden model was ported from TF 1 and does not represent idiomatic Keras code. Our goal is to get to a model written from scratch in idiomatic Keras, which would look very different from the Model Garden example.

Exactly. A while ago I was trying to implement RetinaNet. I looked at the model garden code for the same,  most of which, is taken from the TF Object Detection API. IMO, TFOD was written in a way so that they can add more and more models with minimal changes in `matcher`, `target_assigner`, etc. Although for that use case it works really well, in terms of `flexibility`, `simplicity` and `readability`, it is not the code many would be interested to look into

**Comment by srihari-humbarwadi** [comment]
@AakashKumarNain yeah that's right, I am working on simplifying my previous attempt on implementing the model, although I have having some issues with convergence on a relatively small dataset like Pascal VOC2012, I believe to have managed to reduce the code complexity by a good amount
